### ISSUE TITLE
add "Timeout" option and retry for function `runKubectl`

### ISSUE THREAD

**Issue by henrywangx** [argoproj/argo-workflows#11965]
# Summary

add "Timeout" option and retry for function `runKubectl`.

## Use Cases

1. If the apiserver is unreachable for some reason, `runKubectl` will hang. We need to set a reasonable timeout and retry after few seconds.

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

**Comment by agilgur5** [comment]
Potentially a duplicate of #7179. You can also workaround this by using `kubectl` directly in a `script` template, for example.

Since #7860 is open, I don't think more will be added to the `resource` spec, preferring a plugin instead, but patches that do not change the spec could be acceptable.

**Comment by agilgur5** [comment]
#5549 also mentions using `activeDeadlineSeconds` as a workaround

**Comment by henrywangx** [comment]
> Potentially a duplicate of #7179. You can also workaround this by using `kubectl` directly in a `script` template, for example.
> 
> Since #7860 is open, I don't think more will be added to the `resource` spec, preferring a plugin instead, but patches that do not change the spec could be acceptable.

No, we use the `resource`, which uses argo controller's `runKubectl` to create/get k8s object. And we faced the hang problem when it try to get object's status.

**Comment by agilgur5** [comment]
Yes #7179 and #7860 are about `resource` templates.

I mentioned a `script` for workaround purposes.

**Comment by terrytangyuan** [comment]
Have you tried the workaround I mentioned in https://github.com/argoproj/argo-workflows/issues/5549#issuecomment-810448427?
